File: dotnet/PerformanceUtils/PerformanceMetrics.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace PerformanceUtils
{
    /// <summary>
    /// Enhanced performance metrics tracker providing accurate memory, CPU, and statistical measurements.
    /// </summary>
    public class PerformanceMetrics
    {
        private readonly Process _process;
        private readonly Stopwatch _stopwatch;
        private readonly List<double> _measurements;
        private readonly List<MemorySnapshot> _memorySnapshots;
        private readonly List<CpuSnapshot> _cpuSnapshots;
        
        private long _startWorkingSet;
        private long _startPrivateBytes;
        private DateTime _lastCpuCheck;
        private TimeSpan _lastCpuTime;
        private int _gcGen0Start;
        private int _gcGen1Start;
        private int _gcGen2Start;

        public PerformanceMetrics()
        {
            _process = Process.GetCurrentProcess();
            _stopwatch = new Stopwatch();
            _measurements = new List<double>();
            _memorySnapshots = new List<MemorySnapshot>();
            _cpuSnapshots = new List<CpuSnapshot>();
        }

        /// <summary>
        /// Start performance measurement session.
        /// </summary>
        public void Start()
        {
            // Force garbage collection to get clean baseline
            GC.Collect();
            GC.WaitForPendingFinalizers();
            GC.Collect();

            _process.Refresh();
            _startWorkingSet = _process.WorkingSet64;
            _startPrivateBytes = _process.PrivateMemorySize64;
            
            _gcGen0Start = GC.CollectionCount(0);
            _gcGen1Start = GC.CollectionCount(1);
            _gcGen2Start = GC.CollectionCount(2);
            
            _lastCpuCheck = DateTime.UtcNow;
            _lastCpuTime = _process.TotalProcessorTime;
            
            _stopwatch.Start();
        }

        /// <summary>
        /// Record a single measurement (e.g., one iteration time in milliseconds).
        /// </summary>
        public void RecordMeasurement(double valueMs)
        {
            _measurements.Add(valueMs);
        }

        /// <summary>
        /// Capture a memory snapshot at the current point in time.
        /// </summary>
        public void CaptureMemorySnapshot()
        {
            _process.Refresh();
            var snapshot = new MemorySnapshot
            {
                TimestampMs = _stopwatch.ElapsedMilliseconds,
                WorkingSetMB = _process.WorkingSet64 / 1024.0 / 1024.0,
                PrivateMemoryMB = _process.PrivateMemorySize64 / 1024.0 / 1024.0,
                ManagedMemoryMB = GC.GetTotalMemory(false) / 1024.0 / 1024.0,
                Gen0Collections = GC.CollectionCount(0) - _gcGen0Start,
                Gen1Collections = GC.CollectionCount(1) - _gcGen1Start,
                Gen2Collections = GC.CollectionCount(2) - _gcGen2Start
            };
            _memorySnapshots.Add(snapshot);
        }

        /// <summary>
        /// Capture a CPU usage snapshot.
        /// </summary>
        public void CaptureCpuSnapshot()
        {
            _process.Refresh();
            var currentTime = DateTime.UtcNow;
            var currentCpuTime = _process.TotalProcessorTime;
            
            var wallTimeDiff = (currentTime - _lastCpuCheck).TotalMilliseconds;
            var cpuTimeDiff = (currentCpuTime - _lastCpuTime).TotalMilliseconds;
            
            if (wallTimeDiff > 0)
            {
                var cpuPercent = (cpuTimeDiff / wallTimeDiff / Environment.ProcessorCount) * 100.0;
                
                var snapshot = new CpuSnapshot
                {
                    TimestampMs = _stopwatch.ElapsedMilliseconds,
                    CpuPercent = cpuPercent,
                    ThreadCount = _process.Threads.Count
                };
                _cpuSnapshots.Add(snapshot);
            }
            
            _lastCpuCheck = currentTime;
            _lastCpuTime = currentCpuTime;
        }

        /// <summary>
        /// Stop measurement and calculate comprehensive metrics.
        /// </summary>
        public MetricsResult GetResult()
        {
            _stopwatch.Stop();
            _process.Refresh();

            var result = new MetricsResult
            {
                TotalElapsedMs = _stopwatch.ElapsedMilliseconds,
                MeasurementCount = _measurements.Count
            };

            // Statistical measurements
            if (_measurements.Any())
            {
                var sorted = _measurements.OrderBy(x => x).ToList();
                result.Mean = _measurements.Average();
                result.Min = sorted.First();
                result.Max = sorted.Last();
                result.Median = GetPercentile(sorted, 0.5);
                result.P90 = GetPercentile(sorted, 0.90);
                result.P95 = GetPercentile(sorted, 0.95);
                result.P99 = GetPercentile(sorted, 0.99);
                result.StandardDeviation = CalculateStdDev(_measurements, result.Mean);
            }

            // Memory metrics
            result.WorkingSetDeltaMB = (_process.WorkingSet64 - _startWorkingSet) / 1024.0 / 1024.0;
            result.PrivateMemoryDeltaMB = (_process.PrivateMemorySize64 - _startPrivateBytes) / 1024.0 / 1024.0;
            result.PeakWorkingSetMB = _process.PeakWorkingSet64 / 1024.0 / 1024.0;
            result.PeakPrivateMemoryMB = _process.PeakPagedMemorySize64 / 1024.0 / 1024.0;

            // GC metrics
            result.Gen0Collections = GC.CollectionCount(0) - _gcGen0Start;
            result.Gen1Collections = GC.CollectionCount(1) - _gcGen1Start;
            result.Gen2Collections = GC.CollectionCount(2) - _gcGen2Start;
            result.TotalGCPauseTimeMs = GC.GetTotalPauseDuration().TotalMilliseconds;

            // CPU metrics
            if (_cpuSnapshots.Any())
            {
                result.AverageCpuPercent = _cpuSnapshots.Average(s => s.CpuPercent);
                result.MaxCpuPercent = _cpuSnapshots.Max(s => s.CpuPercent);
            }

            // Detailed snapshots
            result.MemorySnapshots = _memorySnapshots;
            result.CpuSnapshots = _cpuSnapshots;

            return result;
        }

        private static double GetPercentile(List<double> sortedValues, double percentile)
        {
            if (sortedValues.Count == 0) return 0;
            
            var index = percentile * (sortedValues.Count - 1);
            var lower = (int)Math.Floor(index);
            var upper = (int)Math.Ceiling(index);
            
            if (lower == upper) return sortedValues[lower];
            
            var fraction = index - lower;
            return sortedValues[lower] * (1 - fraction) + sortedValues[upper] * fraction;
        }

        private static double CalculateStdDev(List<double> values, double mean)
        {
            if (values.Count <= 1) return 0;
            
            var sumOfSquares = values.Sum(val => Math.Pow(val - mean, 2));
            return Math.Sqrt(sumOfSquares / (values.Count - 1));
        }
    }

    public class MemorySnapshot
    {
        public long TimestampMs { get; set; }
        public double WorkingSetMB { get; set; }
        public double PrivateMemoryMB { get; set; }
        public double ManagedMemoryMB { get; set; }
        public int Gen0Collections { get; set; }
        public int Gen1Collections { get; set; }
        public int Gen2Collections { get; set; }
    }

    public class CpuSnapshot
    {
        public long TimestampMs { get; set; }
        public double CpuPercent { get; set; }
        public int ThreadCount { get; set; }
    }

    public class MetricsResult
    {
        public long TotalElapsedMs { get; set; }
        public int MeasurementCount { get; set; }
        
        // Timing statistics
        public double Mean { get; set; }
        public double Median { get; set; }
        public double Min { get; set; }
        public double Max { get; set; }
        public double P90 { get; set; }
        public double P95 { get; set; }
        public double P99 { get; set; }
        public double StandardDeviation { get; set; }
        
        // Memory metrics
        public double WorkingSetDeltaMB { get; set; }
        public double PrivateMemoryDeltaMB { get; set; }
        public double PeakWorkingSetMB { get; set; }
        public double PeakPrivateMemoryMB { get; set; }
        
        // GC metrics
        public int Gen0Collections { get; set; }
        public int Gen1Collections { get; set; }
        public int Gen2Collections { get; set; }
        public double TotalGCPauseTimeMs { get; set; }
        
        // CPU metrics
        public double AverageCpuPercent { get; set; }
        public double MaxCpuPercent { get; set; }
        
        // Detailed snapshots
        public List<MemorySnapshot> MemorySnapshots { get; set; } = new();
        public List<CpuSnapshot> CpuSnapshots { get; set; } = new();
    }
}

File: python/performance_utils/performance_metrics.py
"""
Enhanced performance metrics tracker for Python providing accurate memory, CPU, and statistical measurements.
"""

import gc
import os
import time
from dataclasses import dataclass, field
from typing import List
import psutil
import statistics


@dataclass
class MemorySnapshot:
    """Memory usage snapshot at a point in time."""
    timestamp_ms: float
    rss_mb: float
    vms_mb: float
    shared_mb: float
    available_system_mb: float
    gc_gen0_count: int
    gc_gen1_count: int
    gc_gen2_count: int


@dataclass
class CpuSnapshot:
    """CPU usage snapshot at a point in time."""
    timestamp_ms: float
    cpu_percent: float
    thread_count: int


@dataclass
class MetricsResult:
    """Complete performance metrics result."""
    total_elapsed_ms: float
    measurement_count: int
    
    # Timing statistics
    mean: float = 0.0
    median: float = 0.0
    min: float = 0.0
    max: float = 0.0
    p90: float = 0.0
    p95: float = 0.0
    p99: float = 0.0
    stdev: float = 0.0
    
    # Memory metrics
    rss_delta_mb: float = 0.0
    vms_delta_mb: float = 0.0
    peak_rss_mb: float = 0.0
    peak_vms_mb: float = 0.0
    
    # GC metrics
    gc_gen0_collections: int = 0
    gc_gen1_collections: int = 0
    gc_gen2_collections: int = 0
    
    # CPU metrics
    average_cpu_percent: float = 0.0
    max_cpu_percent: float = 0.0
    
    # Detailed snapshots
    memory_snapshots: List[MemorySnapshot] = field(default_factory=list)
    cpu_snapshots: List[CpuSnapshot] = field(default_factory=list)


class PerformanceMetrics:
    """Enhanced performance metrics tracker."""
    
    def __init__(self):
        self._process = psutil.Process(os.getpid())
        self._start_time = None
        self._measurements: List[float] = []
        self._memory_snapshots: List[MemorySnapshot] = []
        self._cpu_snapshots: List[CpuSnapshot] = []
        
        self._start_rss = 0
        self._start_vms = 0
        self._gc_gen0_start = 0
        self._gc_gen1_start = 0
        self._gc_gen2_start = 0
    
    def start(self):
        """Start performance measurement session."""
        # Force garbage collection to get clean baseline
        gc.collect()
        gc.collect()
        gc.collect()
        
        # Get starting memory
        mem_info = self._process.memory_info()
        self._start_rss = mem_info.rss
        self._start_vms = mem_info.vms
        
        # Get starting GC counts
        gc_counts = gc.get_count()
        self._gc_gen0_start = gc_counts[0]
        self._gc_gen1_start = gc_counts[1]
        self._gc_gen2_start = gc_counts[2]
        
        self._start_time = time.perf_counter()
    
    def record_measurement(self, value_ms: float):
        """Record a single measurement (e.g., one iteration time in milliseconds)."""
        self._measurements.append(value_ms)
    
    def capture_memory_snapshot(self):
        """Capture a memory snapshot at the current point in time."""
        elapsed_ms = (time.perf_counter() - self._start_time) * 1000
        mem_info = self._process.memory_info()
        
        # Get GC stats
        gc_counts = gc.get_count()
        
        # Get system memory
        vm = psutil.virtual_memory()
        
        snapshot = MemorySnapshot(
            timestamp_ms=elapsed_ms,
            rss_mb=mem_info.rss / 1024 / 1024,
            vms_mb=mem_info.vms / 1024 / 1024,
            shared_mb=getattr(mem_info, 'shared', 0) / 1024 / 1024,
            available_system_mb=vm.available / 1024 / 1024,
            gc_gen0_count=gc_counts[0] - self._gc_gen0_start,
            gc_gen1_count=gc_counts[1] - self._gc_gen1_start,
            gc_gen2_count=gc_counts[2] - self._gc_gen2_start
        )
        self._memory_snapshots.append(snapshot)
    
    def capture_cpu_snapshot(self):
        """Capture a CPU usage snapshot."""
        elapsed_ms = (time.perf_counter() - self._start_time) * 1000
        
        # Get CPU percent with a short interval
        cpu_percent = self._process.cpu_percent(interval=0.1)
        thread_count = self._process.num_threads()
        
        snapshot = CpuSnapshot(
            timestamp_ms=elapsed_ms,
            cpu_percent=cpu_percent,
            thread_count=thread_count
        )
        self._cpu_snapshots.append(snapshot)
    
    def get_result(self) -> MetricsResult:
        """Stop measurement and calculate comprehensive metrics."""
        total_elapsed_ms = (time.perf_counter() - self._start_time) * 1000
        
        result = MetricsResult(
            total_elapsed_ms=total_elapsed_ms,
            measurement_count=len(self._measurements)
        )
        
        # Statistical measurements
        if self._measurements:
            sorted_measurements = sorted(self._measurements)
            result.mean = statistics.mean(self._measurements)
            result.min = min(self._measurements)
            result.max = max(self._measurements)
            result.median = statistics.median(self._measurements)
            result.p90 = self._percentile(sorted_measurements, 0.90)
            result.p95 = self._percentile(sorted_measurements, 0.95)
            result.p99 = self._percentile(sorted_measurements, 0.99)
            if len(self._measurements) > 1:
                result.stdev = statistics.stdev(self._measurements)
        
        # Memory metrics
        mem_info = self._process.memory_info()
        result.rss_delta_mb = (mem_info.rss - self._start_rss) / 1024 / 1024
        result.vms_delta_mb = (mem_info.vms - self._start_vms) / 1024 / 1024
        
        # Peak memory (if available)
        try:
            mem_full = self._process.memory_full_info()
            result.peak_rss_mb = getattr(mem_full, 'uss', mem_info.rss) / 1024 / 1024
            result.peak_vms_mb = mem_info.vms / 1024 / 1024
        except:
            result.peak_rss_mb = mem_info.rss / 1024 / 1024
            result.peak_vms_mb = mem_info.vms / 1024 / 1024
        
        # GC metrics
        gc_counts = gc.get_count()
        result.gc_gen0_collections = gc_counts[0] - self._gc_gen0_start
        result.gc_gen1_collections = gc_counts[1] - self._gc_gen1_start
        result.gc_gen2_collections = gc_counts[2] - self._gc_gen2_start
        
        # CPU metrics
        if self._cpu_snapshots:
            result.average_cpu_percent = statistics.mean(s.cpu_percent for s in self._cpu_snapshots)
            result.max_cpu_percent = max(s.cpu_percent for s in self._cpu_snapshots)
        
        # Detailed snapshots
        result.memory_snapshots = self._memory_snapshots
        result.cpu_snapshots = self._cpu_snapshots
        
        return result
    
    @staticmethod
    def _percentile(sorted_values: List[float], percentile: float) -> float:
        """Calculate percentile from sorted values."""
        if not sorted_values:
            return 0.0
        
        index = percentile * (len(sorted_values) - 1)
        lower = int(index)
        upper = int(index + 0.5) if index != lower else lower
        
        if lower == upper or upper >= len(sorted_values):
            return sorted_values[lower]
        
        fraction = index - lower
        return sorted_values[lower] * (1 - fraction) + sorted_values[upper] * fraction
